src/ecos.c

```diff
@@ -1652,25 +1652,42 @@ void ECOS_updateData(pwork *w, pfloat *Gpr, pfloat *Apr,
     idxint k;
 
 #if defined EQUILIBRATE && EQUILIBRATE > 0
+    /* only unset the equilibration if new data is outside old data */
+    if (
+        ((Gpr + w->G->nnz < w->G->pr) || (w->G->pr + w->G->nnz < Gpr)) &&
+        ((Apr + w->A->nnz < w->A->pr) || (w->A->pr + w->A->nnz < Apr)) &&
+        ((c + w->n < w->c) || (w->c + w->n < c)) &&
+        ((h + w->m < w->h) || (w->h + w->m < h)) &&
+        ((b + w->p < w->b) || (w->b + w->p < b))
+    ){
     unset_equilibration(w);
+    }
 #endif
 
     /* update pointers */
-    w->G->pr = Gpr;
-    w->A->pr = Apr;
+    if(w->G){
+        w->G->pr = Gpr;
+        w->h = h;
+    }
+    if(w->A){
+        w->A->pr = Apr;
+        w->b = b;
+    }
     w->c = c;
-    w->h = h;
-    w->b = b;
 
 #if defined EQUILIBRATE && EQUILIBRATE > 0
     set_equilibration(w);
 #endif
```

Comment:
@EmbersArc I've been thinking about this and it seems that this change makes it so that `unset_equilibration` happens  only if the conditions are met, but `set_equilibration` happens unconditionally. This could mean the data gets equilibrated multiple times in a row without a corresponding un-equilibration. 

I'm not that familiar with the equilibration process - have you looked into this more? If what I'm describing is not an issue, maybe adding a comment saying so to the `set_equilibration` call here would help avoid future confusion.

Reply:
Thanks for the review. In my view, the equilibration can be done unconditionally. The new data is always expected to be unequilibrated. The first check only happens to make sure that the old data gets unequilibrated if it did not get overwritten by the new data. Existing code using this function might rely on this step. I added some comments to make this clearer.